server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -19,11 +19,41 @@ def __init__(self, data, graph_method="umap", diffexp_method="ttest"):
         self._add_mandatory_annotations()
         self.cell_count = self.data.shape[0]
         self.gene_count = self.data.shape[1]
+        self._create_schema()
         self.graph_method = graph_method
         self.diffexp_method = diffexp_method
 
-    def _set_cell_names(self):
-        self.data.obs["cell_name"] = list(self.data.obs.index)
+    def _create_schema(self):
+        self.schema = {
+            "dataframe": {
+                "nObs": self.cell_count,
+                "nVar": self.gene_count,
+                "type": str(self.data.X.dtype)
+            },
+            "annotations": {
+                "obs": [],
+                "var": []
+            }
+        }
+        for ax in Axis:
+            curr_axis = getattr(self.data, str(ax))
+            for ann in curr_axis:
+                ann_schema = {"name": ann}
+                data_kind = curr_axis[ann].dtype.kind
+                if data_kind == 'f':
+                    ann_schema["type"] = "float32"
+                elif data_kind in ['i', 'u']:
+                    ann_schema["type"] = "int32"
+                elif data_kind == "?":
+                    ann_schema["type"] = "boolean"
+                elif data_kind == "O" and curr_axis[ann].dtype == "object":
+                    ann_schema["type"] = "string"
```

Comment:
Are unicode (kind is 'U') strings missing?  Or does scanpy not use them?

Reply:
Pandas doesn't use them; strings turn into objects. http://pbpython.com/pandas_dtypes.html